README.md
This package parses and handles LRC strings, along with handy methods for them.

## Features
* It can parse LRC strings or create new ones from scratch
* Simple and intuitive
* Adds handy extension methods for strings specific to LRCs
* Ability to stream the lyrics at the set durations

## Usage
### Installing
Installation instructions can be found on the "Installing" tab on pub.dev.

### Importing
As always, import first this package:
```dart
import 'package:lrc/lrc.dart';
```

### Parsing
This package only accepts strings so that this can be used for Dart JS and for the web. To accept from files, parse and extract the contents of the file yourself and put it into a string.

You can parse strings using the following:

```dart
//this string contains a LRC that can be parsed
String unparsedLrc = """
[ti:Never Gonna Give You Up]
[ar:Rick Astley]
[la:en]
[00:18.78]We're no strangers to love
[00:22.83]You know the rules and so do I
[00:27.11]A full commitment's what I'm thinking of
[00:31.25]You wouldn't get this from any other guy
"""; 

//parse the string using the static method
Lrc parsedLrc = Lrc.parse(unparsedLrc);

//parse the string using the handy extension method
Lrc parsedLrc = unparsedLrc.toLrc();
```

Parsing will return a `Lrc` object if successful. If the string is not a valid LRC, then a `FormatExeption` will be thrown.

You can check if the string is valid using the following:

```dart
//we don't know if this is valid lrc
String idk = ...; 

//check using the static method
bool isValid = Lrc.isValid(idk);

//check using the handy extension getter
bool isValid = idk.isValidLrc;
```
Checking will return a `bool`.

### Creating
If you want to create a LRC from scratch, you can do that by creating a new `Lrc` object with its parameters.

```dart
Lrc(
  type: //[LrcTypes] The type of the lrc. See below for more details. (required)
  lyrics: //[List<LrcLine>] The lyrics themselves. See below for more details. (required)
  artist: //[String] The artist of the song (optional)
  album: //[String] The album of the song (optional)
  title: //[String] The title of the song (optional)
  author: //[String] The author of the lyrics (optional)
  creator: //[String] The creator of the LRC file (optional)
  program: //[String] The program that created the LRC file (optional)
  version: //[String] The version of the program that created the LRC file (optional)
  length: //[String] The length/duration of the song  (optional)
  language: //[String] The IETF BCP 47 language tag of the song (optional)
  offset: //[int] The offset of the timing in milliseconds (optional)
)
```

#### Types
LRC comes with three types, simple, extended, and enhanced. See the [Wikipedia article](https://en.wikipedia.org/wiki/LRC_(file_format)) on LRCs for more information on these types. In this package, the types are stored in an enum class called `LrcTypes`.

```dart
LrcTypes.simple
LrcTypes.extended
LrcTypes.enhanced
LrcTypes.extended_enhanced //some lines are extended while some are enhanced
```

#### Lyrics
The lyrics are encoded in a `List<LrcLine>`, which is basically a list of `LrcLine`, each has the following properties:

```dart
LrcLine(
  timestamp: //[Duration] the timestamp wherein this lyric will show (required)
  lyrics: //[String] the actual raw lyrics in this line (required)
  type: //[LrcTypes] the type of LRC for this specific line only
  args: //[Map<String, Object>] addition arguments if the type is not simple (optional)
)
```

If the LRC is extended, then that particular line will have these as the `args`:

```dart
<String, Object>{
  'letter': //[String] a string containing the letter modifier
  'lyrics': //[String] a string containing the lyrics for that line
}
```

If the LRC is enhanced, then that line will have these as the `args`:

```dart
//The key string is the timestamp in angle brackets as seen in the raw lyrics.
//There will be a key-value pair for each timestamp-lyric pair in the raw line.
<String, Object>{
  '<00:00.00>': <String, Object>{
    'duration': //[Duration] the timestamp of the lyrics
	'lyrics': //[String] the lyrics for that timestamp
  }
}
```

### Formatting
To format the stored `Lrc` into an actual string that can be saved into a file later on:

```dart
Lrc parsedLrc = ...; //the Lrc object to format

//format using the format() method
String raw = parsedLrc.format();
```

#### Formatting a single line
To format just a single `LrcLine`:
```dart
LrcLine line = ...; //the LrcLine object to format

//format using the `formattedLine` property
String raw = line.formattedLine;
```

### Streaming
You can stream each lyric at their set duration. To do this:

```dart
streamLrc(Lrc lrc) async {
  // subscribe using the `stream` property
  await for (LrcStream i in lrc.stream) {
    ...
  }

  // subscribe using the handy `toStream()` extension method on List<LrcLine>
  await for (LrcStream i in lrc.lyrics.toStream()) {
    ...
  }
}

//You can also do this with a List<LrcLine>
async streamLrc(List<LrcLine> lines){
	// subscribe using the handy `toStream()` extension method
  await for (LrcStream i in lines.toStream()){
	  ...
  }
}
```
The `LrcStream` class that outputs from each yielding of a stream stores the following:
```dart
LrcStream(
  previous: //[LrcLine] the previous line. Is null if its the first line
  current: //[LrcLine] the current line
  next: //[LrcLine] the next line. Is null if its the last line
  duration: //[Duration] the duration from the current to the next. Is null if its the last line
  position: //[int] the position of the current line
  length: //[int] the number of lines
)
```

## License
This package is licensed under the 3 clause BSD license.
```
Copyright 2023 Yivan000

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this
list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice,
this list of conditions and the following disclaimer in the documentation
and/or other materials provided with the distribution.

3. Neither the name of the copyright holder nor the names of its contributors
may be used to endorse or promote products derived from this software without
specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
```

## UwU
o(〃＾▽＾〃)o

CHANGELOG.md
## 1.0.0

- Initial version.

## 1.0.1

- Fixed a bug that causes `Lrc.format` to return an invalid LRC string [[#3](https://github.com/Yivan000/lrc/pull/3)]
- Added more and fixed some documentation

## 1.0.2

- Fixed a bug where the hundreds of a second time is being saved as thousands of a second [[#4](https://github.com/Yivan000/lrc/pull/4)]

lib/src/lrc_main.dart
/// The parsed LRC class.
///
/// You can instantiate this class directly
/// or parse a string using `Lrc.parse()`.
class Lrc {
  /// The overall type of LRC for this object
  LrcTypes type;

  /// The name of the artist of the song (optional)
  ///
  /// This corresponds to the ID tag `[ar:]`.
  String? artist;

  /// The name of the album of the song (optional)
  ///
  /// This corresponds to the ID tag `[al:]`.
  String? album;

  /// The title of the song (optional)
  ///
  /// This corresponds to the ID tag `[ti:]`.
  String? title;

  /// The name of the author of the lyrics (optional)
  ///
  /// This corresponds to the ID tag `[au:]`.
  String? author;

  /// The name of the creator of the LRC file (optional)
  ///
  /// This corresponds to the ID tag `[by:]`.
  String? creator;

  /// The name of the program that created the LRC file (optional)
  ///
  /// This corresponds to the ID tag `[re:]`.
  String? program;

  /// The version of the program that created the LRC file (optional)
  ///
  /// This corresponds to the ID tag `[ve:]`.
  String? version;

  /// The length of the song (optional)
  ///
  /// This corresponds to the ID tag `[length:]`.
  String? length;

  /// The language of the song, using an IETF BCP 47 language tag (optional)
  ///
  /// This corresponds to the ID tag `[la:]`.
  String? language;

  /// Offset of time in milliseconds, can be positive [shifts time up]
  /// or negative [shifts time down] (optional)
  ///
  /// This corresponds to the ID tag `[offset:]`.
  int? offset;

  /// The list of lyric lines
  List<LrcLine> lyrics;

  /// Handy parameter to get a stream of the lyrics.
  /// See `List<LrcLine>.toStream()`.
  Stream<LrcStream> get stream => lyrics.toStream();

  /// Use this constructor if you want to manually create an LRC from scratch.
  /// Otherwise, parse an LRC string using [Lrc.parse].
  Lrc({
    this.type = LrcTypes.simple,
    required this.lyrics,
    this.artist,
    this.album,
    this.title,
    this.creator,
    this.author,
    this.program,
    this.version,
    this.length,
    this.offset,
    this.language,
  });

  /// Format the lrc to a readable string that can then be
  /// outputted to an LRC file.
  String format() {
    var output = '';

    output += (artist != null) ? '[ar:$artist]\n' : '';
    output += (album != null) ? '[al:$album]\n' : '';
    output += (title != null) ? '[ti:$title]\n' : '';
    output += (length != null) ? '[length:$length]\n' : '';
    output += (creator != null) ? '[by:$creator]\n' : '';
    output += (author != null) ? '[au:$author]\n' : '';
    output += (offset != null) ? '[offset:${offset.toString()}]\n' : '';
    output += (program != null) ? '[re:$program]\n' : '';
    output += (version != null) ? '[ve:$version]\n' : '';
    output += (language != null) ? '[la:$language]\n' : '';

    for (var lyric in lyrics) {
      output += lyric.formattedLine + '\n';
    }

    return output;
  }

  /// Parses an LRC from a string. Throws a `FormatExeption`
  /// if the inputted string is not valid.
  static Lrc parse(String parsed) {
    parsed = parsed.trim();

    if (!isValid(parsed)) {
      throw FormatException('The inputted string is not a valid LRC file');
    }

    // split string into lines, code from Linesplitter().convert(data)
    var lines = ((data) {
      var lines = <String>[];
      var end = data.length;
      var sliceStart = 0;
      var char = 0;
      for (var i = 0; i < end; i++) {
        var previousChar = char;
        char = data.codeUnitAt(i);
        if (char != 13) {
          if (char != 10) continue;
          if (previousChar == 13) {
            sliceStart = i + 1;
            continue;
          }
        }
        lines.add(data.substring(sliceStart, i));
        sliceStart = i + 1;
      }
      if (sliceStart < end) lines.add(data.substring(sliceStart, end));
      return lines;
    })(parsed);

    // temporary storer variables
    String? artist,
        album,
        title,
        length,
        author,
        creator,
        offset,
        program,
        version,
        language;
    LrcTypes? type;
    var lyrics = <LrcLine>[];

    String? setIfMatchTag(String toMatch, String tag) =>
        (RegExp(r'^\[' + tag + r':.*\]$').hasMatch(toMatch))
            ? toMatch.substring(tag.length + 2, toMatch.length - 1).trim()
            : null;

    // loop thru each lines
    for (var i in lines) {
      artist = artist ?? setIfMatchTag(i, 'ar');
      album = album ?? setIfMatchTag(i, 'al');
      title = title ?? setIfMatchTag(i, 'ti');
      author = author ?? setIfMatchTag(i, 'au');
      length = length ?? setIfMatchTag(i, 'length');
      creator = creator ?? setIfMatchTag(i, 'by');
      offset = offset ?? setIfMatchTag(i, 'offset');
      program = program ?? setIfMatchTag(i, 're');
      version = version ?? setIfMatchTag(i, 've');
      language = language ?? setIfMatchTag(i, 'la');

      if (RegExp(r'^\[\d\d:\d\d\.\d\d\].*$').hasMatch(i)) {
        var lyric = i.substring(10).trim();
        var lineType = LrcTypes.simple;
        Map<String, Object>? args;

        // checkers for different types of LRCs
        if (lyric.contains(RegExp(r'^\w:'))) {
          //if extended
          type = (type == LrcTypes.enhanced)
              ? LrcTypes.extended_enhanced
              : LrcTypes.extended;
          args = {
            'letter': lyric[0], // get the letter of the type of person
            'lyrics': lyric.substring(2) // get the rest of the lyrics
          };
          lineType = LrcTypes.extended;
        } else if (lyric.contains(RegExp(r'<\d\d:\d\d\.\d\d>'))) {
          // if enhanced
          type = (type == LrcTypes.extended)
              ? LrcTypes.extended_enhanced
              : LrcTypes.enhanced;
          args = {};
          lineType = LrcTypes.enhanced;
          // for each timestamp in the line, regex has capturing
          // groups to make this easier
          for (var j in RegExp(r'<((\d\d):(\d\d)\.(\d\d))>([^<]+)')
              .allMatches(lyric)) {
            // puts each timestamp+lyrics in the args, no duplicates
            args.putIfAbsent(
              j.group(1)!, //the key is the <mm:ss.xx>
              () => <String, Object>{
                // the value is another map with the duration and lyrics
                'duration': Duration(
                  minutes: int.parse(j.group(2)!),
                  seconds: int.parse(j.group(3)!),
                  milliseconds: int.parse(j.group(4)!) * 10,
                ),
                'lyrics': j.group(5)!.trim()
              },
            );
          }
        }
        final minutes = int.parse(i.substring(1, 3));
        final seconds = int.parse(i.substring(4, 6));
        final hundreds = int.parse(i.substring(7, 9));

        lyrics.add(LrcLine(
          timestamp: Duration(
            minutes: minutes,
            seconds: seconds,
            milliseconds: hundreds * 10,
          ),
          lyrics: lyric,
          type: lineType,
          args: args,
        ));
      }
    }

    return Lrc(
      type: type ?? LrcTypes.simple,
      artist: artist,
      album: album,
      title: title,
      author: author,
      length: length,
      creator: creator,
      offset: (offset != null) ? int.tryParse(offset) : null,
      program: program,
      version: version,
      lyrics: lyrics,
      language: language,
    );
  }

  /// Checks if the string [input] is a valid LRC using Regex.
  static bool isValid(String input) => RegExp(
          r'^([\r\n]*\[((ti)|(a[rlu])|(by)|([rv]e)|(length)|(offset)|(la)):.+\][\r\n]*)*([\r\n]*\[\d\d:\d\d\.\d\d\].*){2,}[\r\n]*$')
      .hasMatch(input.trim());

  @override
  String toString() {
    var lyrics = this.lyrics.join('\n');

    return '''
    Type: '$type'
    Artist: '$artist'
    Album: '$album'
    Title: '$title'
    Author: '$author'
    Creator: '$creator'
    Program: '$program'
    Length: '$length'
    Language: '$language'
    Offset: '$offset'
    Lyrics: '$lyrics'
    ''';
  }
}

///The types of LRC
enum LrcTypes {
  ///A simple LRC, with no extra formatting, etc
  simple,

  ///LRC with modifiers at the start in the form `A: foo`
  extended,

  ///LRC with additional timestamps per line in the form `<00:00.00> foo`
  enhanced,

  ///LRC that some lines are extended and some are enhanced
  extended_enhanced
}

///A line of lyrics, with its defined duration and raw lyrics
class LrcLine {
  ///timestamp for the lyrics wherein it'll be displayed
  Duration timestamp;

  ///the raw lyrics for the line
  String lyrics;

  ///the additional arguments for other lrc types
  Map<String, Object>? args;

  ///the type of lrc for this line
  LrcTypes type;

  LrcLine({
    required this.timestamp,
    required this.lyrics,
    required this.type,
    this.args,
  });

  ///get the string for a formatted line
  String get formattedLine {
    ///function to add leading zeros
    String f(int x) => x.toString().padLeft(2, '0');

    // LRC format doesn't accept hours.
    final minutes = timestamp.inMinutes % 60,
        seconds = timestamp.inSeconds % 60,
        hundreds = timestamp.inMilliseconds % 1000 ~/ 10;

    return '[${f(minutes)}:${f(seconds)}.${f(hundreds)}]$lyrics';
  }

  @override
  String toString() {
    return '''
      Timestamp: '$timestamp'
      Lyrics: '$lyrics'
      Args: '$args'
    ''';
  }
}

/// A data class to store each yielding of the stream
class LrcStream {
  /// The previous line. Is null if the current line is the fist position.
  LrcLine? previous;

  /// Tthe current line
  LrcLine current;

  /// The next line. Is null if the current line is the last position.
  LrcLine? next;

  /// The duration from the current to the next. Is null if the current line is the last position.
  Duration? duration;

  /// The position of the current line
  int position;

  /// The total number of lines in the stream
  int length;

  /// The main constructor for a LrcStream
  LrcStream(
      {this.previous,
      required this.current,
      this.next,
      this.duration,
      required this.position,
      required this.length})
      //position should be greater than or equal to 0
      : assert(position >= 0),
        //the length should be greater than or equal to the position
        assert(length >= position),
        //previous is null only if position is 0
        assert((previous == null) ? position == 0 : true),
        //next is null only if position is the last
        assert((next == null) ? position == length : true);
}

/// Handy extensions on lists of LrcLine
extension LrcLineExtensions on List<LrcLine> {
  /// Creates a stream for each lyric using their durations
  Stream<LrcStream> toStream() async* {
    for (var i = 0; i < length; i++) {
      var lineCurrent = this[i];
      var lineNext = (i + 1 < length) ? this[i + 1] : null;
      var durationToNext = (lineNext != null)
          ? Duration(
              milliseconds: lineNext.timestamp.inMilliseconds -
                  lineCurrent.timestamp.inMilliseconds)
          : null;
      yield LrcStream(
          duration: durationToNext,
          previous: (i != 0) ? this[i - 1] : null,
          current: lineCurrent,
          next: lineNext,
          position: i,
          length: length - 1);
      if (durationToNext != null) {
        await Future.delayed(durationToNext);
      }
    }
  }
}

/// Handy extensions on strings
extension StringExtensions on String {
  /// Handy extension method that parses the string to an [Lrc]
  Lrc toLrc() => Lrc.parse(this);

  /// Handy extension getter if the given string is a valid LRC
  bool get isValidLrc => Lrc.isValid(this);
}
